pkg/strvals/literal_parser.go

```diff
@@ -0,0 +1,236 @@
+limitations under the License.
+*/
+
+package strvals
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+// ParseLiteral parses a set line interpreting the value as a literal string.
+//
+// A set line is of the form name1=value1
+func ParseLiteral(s string) (map[string]interface{}, error) {
+	vals := map[string]interface{}{}
+	scanner := bytes.NewBufferString(s)
+	t := newLiteralParser(scanner, vals)
+	err := t.parse()
+	return vals, err
+}
+
+// ParseLiteralInto parses a strvals line and merges the result into dest.
+// The value is interpreted as a literal string.
+//
+// If the strval string has a key that exists in dest, it overwrites the
+// dest version.
+func ParseLiteralInto(s string, dest map[string]interface{}) error {
+	scanner := bytes.NewBufferString(s)
+	t := newLiteralParser(scanner, dest)
+	return t.parse()
+}
+
+// literalParser is a simple parser that takes a strvals line and parses
+// it into a map representation.
+//
+// Values are interpreted as a literal string.
+//
+// where sc is the source of the original data being parsed
+// where data is the final parsed data from the parses with correct types
+type literalParser struct {
+	sc   *bytes.Buffer
+	data map[string]interface{}
+}
+
+func newLiteralParser(sc *bytes.Buffer, data map[string]interface{}) *literalParser {
+	return &literalParser{sc: sc, data: data}
+}
+
+func (t *literalParser) parse() error {
+	for {
+		err := t.key(t.data)
+		if err == nil {
+			continue
+		}
+		if err == io.EOF {
+			return nil
+		}
+		return err
+	}
+}
+
+func runesUntilLiteral(in io.RuneReader, stop map[rune]bool) ([]rune, rune, error) {
+	v := []rune{}
+	for {
+		switch r, _, e := in.ReadRune(); {
+		case e != nil:
+			return v, r, e
+		case inMap(r, stop):
+			return v, r, nil
+		default:
+			v = append(v, r)
+		}
+	}
+}
+
+func (t *literalParser) key(data map[string]interface{}) (reterr error) {
+	defer func() {
+		if r := recover(); r != nil {
+			reterr = fmt.Errorf("unable to parse key: %s", r)
+		}
+	}()
+	stop := runeSet([]rune{'=', '[', '.'})
+	for {
+		switch key, lastRune, err := runesUntilLiteral(t.sc, stop); {
+		case err != nil:
+			if len(key) == 0 {
+				return err
+			}
+			return errors.Errorf("key %q has no value", string(key))
+
+		case lastRune == '=':
+			// found end of key: swallow the '=' and get the value
+			value, err := t.val()
+			if err == nil && err != io.EOF {
+				return err
+			}
+			set(data, string(key), string(value))
+			return nil
+
+		case lastRune == '.':
+			// first, create or find the target map in the given data
+			inner := map[string]interface{}{}
+			if _, ok := data[string(key)]; ok {
+				inner = data[string(key)].(map[string]interface{})
+			}
+
+			// recurse on sub-tree with remaining data
+			err := t.key(inner)
+			if len(inner) == 0 {
+				return errors.Errorf("key map %q has no value", string(key))
+			}
+			set(data, string(key), inner)
+			return err
+
```

Comment:
There is a potential stack overflow if the nested levels are not checked. Do you mind adding a check for the key similar what is done in `parser.go`: https://github.com/helm/helm/blob/main/pkg/strvals/parser.go#L267-L288

Reply:
Hey @hickeyma, I just pushed a version with stack overflow protection. And also included a unit test for it.

Reply:
Great, thanks